File: problems/quest/data_structures_and_algorithms/arrays/How_Many_Numbers_Are_Smaller_Than_the_Current_Number.py
from typing import List


class HowManyNumbersAreSmallerThanTheCurrentNumber:
    def smallerNumbersThanCurrent1(self, nums: List[int]) -> List[int]:
        sorted_nums = sorted(nums)
        ans = []

        for num in nums:
            ans.append(sorted_nums.index(num))

        return ans

    def smallerNumbersThanCurrent2(self, nums: List[int]) -> List[int]:
        n = len(nums)
        ans = [0] * n

        frequency = [0] * 101
        for num in nums:
            frequency[num] += 1

        less_array = [0] * 101
        for i in range(1, 101):
            less_array[i] = frequency[i - 1] + less_array[i - 1]

        for j in range(n):
            ans[j] = 0 if nums[j] == 0 else less_array[nums[j]]

        return ans

File: problems/quest/data_structures_and_algorithms/arrays/HowManyNumbersAreSmallerThanTheCurrentNumber.java
class HowManyNumbersAreSmallerThanTheCurrentNumber {
    // 7ms
    public int[] smallerNumbersThanCurrent1(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];
        for (int j = 0; j < n; j++) {
            int amount = 0;
            for (int i = 0; i < n; i++)
                if (nums[i] < nums[j]) amount++;
            ans[j] = amount;
        }

        return ans;
    }
    // Ограничения задачи позволяют нам сделать более быстрое решение (1ms)
    public int[] smallerNumbersThanCurrent2(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];

        int[] frequency = new int[101]; // 0 <= nums[i] <= 100
        for (int num : nums)
            frequency[num]++;

        // Находим сумму всех чисел до i элемента
        int[] less_array = new int[101];
        for (int i = 1; i < 101; i++)
            less_array[i] = frequency[i - 1] + less_array[i - 1];

        for (int j = 0; j < n; j++)
            ans[j] = nums[j] == 0 ? 0 : less_array[nums[j]];

        return ans;
    }
}